majilove/models.py

```diff
@@ -106,3 +108,402 @@ def display_blood_type(self):
+            'template': _(u'Cut-in Bonus Score +{percentage}%'),
+            'japanese_template': u'カットインボーナスのスコア{percentage}%上昇'
+        }),
+        (4, {
+            'translation': _(u'Good lock'),
+            'english': 'Good lock',
+            # need someone to check this, feels too long
+            'japanese_translation': u'BADをGREATに',
+            'icon': 'perfectlock',
+
+            'variables': ['note_count'],
+            'template': _(u'Bad > Good ({note_count} Times)'),
+            'japanese_template': _(u'BADを{note_count}回GREATにする')
+        }),
+        (5, {
+            'translation': _(u'Great lock'),
+            'english': 'Great lock',
+            # also seems long, need to check which comma to use
+            'japanese_translation': u'BAD,GREATをPERFECTに',
+            'icon': 'perfectlock',
+
+            'variables': ['note_count'],
+            'template': _(u'Bad/Good > Great ({note_count} Times)'),
+            'japanese_template': u'BAD,GREATを{note_count}回PERFECTにする'
+        }),
+        (6, {
+            'translation': _(u'Healer'),
+            'english': 'Healer',
+            'japanese_translation': u'ライフ回復ノーツ',
+            'icon': 'healer',
+
+            'variables': ['note_count'],
+            'template': _(u'Stamina Recovery Notes +{note_count}'),
+            'japanese_template': u'ライフ回復ノーツを{note_count}個追加'
+        })
+    ])
+
+    ALL_VARIABLES = { item: True for sublist in [ _info['variables'] for _info in SKILL_TYPES.values() ] for item in sublist }.keys()
+
+    SKILL_TYPE_WITHOUT_I_CHOICES = True
+    SKILL_TYPE_CHOICES = [(_name, _info['translation']) for _name, _info in SKILL_TYPES.items()]
+
+    i_skill_type = models.PositiveIntegerField(_('Skill'), choices=SKILL_TYPE_CHOICES, null=True, db_index=True)
+    japanese_skill_type = property(getInfoFromChoices('skill_type', SKILL_TYPES, 'japanese_translation'))
+    skill_icon = property(getInfoFromChoices('skill_type', SKILL_TYPES, 'icon'))
+
+    @property
+    def skill_template(self):
+        return self.SKILL_TYPES[self.skill_type]['template']
+
+    @property
+    def japanese_skill_template(self):
+        return self.SKILL_TYPES[self.skill_type]['japanese_template']
+
+    @property
+    def skill_variables(self):
+        return {
+            key: getattr(self, u'skill_{}'.format(key))
+            for key in self.SKILL_TYPES[self.skill_type]['variables']
+        }
+
+    @property
+    def skill(self):
+        if self.i_skill_type is None: return None
+        return self.skill_template.format(**self.skill_variables)
+
+    @property
+    def japanese_skill(self):
+        if self.i_skill_type is None: return None
+        return self.japanese_skill_template.format(**self.skill_variables)
+
+    skill_note_count = models.PositiveIntegerField('{note_count}', null=True)
+    # should percentage be split into different variales for perfect score and cutin?
+    skill_percentage = models.FloatField('{percentage}', null=True)
```

Comment:
I'm not sure how skill percentage changes, but if it's linear with a set step, you can probably calculate it automatically (that's what we do for bandori).

If it's linear between 2 values or if all values per level are different we'll need to store the value differently. This should be done later when collectibles are up though, since changing the value of the skill depending on the level makes sense in the context of collectibles, when a user stored the level.

Reply:
It's linear, but there's two very different uses of the percentage: the cut-in skill starts at 10%/20%/30% (depending on rarity) and increases by 10% each level. The perfect score skill starts at 2.5%/2.8% and increase 0.3% each level. That's why I have the comment about possibly splitting the variable into separate fields. (The various note based skills start at 1/2/3/4 notes and increase by one note each level). I'm assuming that calculating the specific value at a certain skill level would be done in the collectible version of the model.